santorini-common/src/command.rs

```diff
@@ -18,28 +16,83 @@ pub struct MovementCommand {
-        let worker = state
-            .mut_space(&self.from)
-            .mut_worker()
-            .take()
-            .ok_or_else(|| {
-                SantoriniError::InvalidArgument(format!("There is no worker at {}", self.from,))
-            })?;
+    fn can_execute(&self, state: &State) -> Result<(), SantoriniError> {
+        // A move can execute if
+
+        // 1) the `from` space has a worker
+        let worker = match state.space(&self.from).worker() {
+            Some(worker) => worker,
+            None => {
+                return Err(SantoriniError::InvalidArgument(format!(
+                    "There is no worker at {}",
+                    self.from
+                )));
+            }
+        };
+
+        // 2) the worker belongs to the current player's turn
         if worker.player() != state.current_player() {
             return Err(SantoriniError::InvalidArgument(format!(
                 "Current player {} cannot control {} worker.",
                 state.current_player(),
                 worker.player()
             )));
         }
-        let previous = state.mut_space(&self.to).mut_worker().replace(worker);
-        if previous.is_some() {
+
+        // 3) it is a move phase
+        // TODO
+
+        // 4) it is a valid move in one of the 8 directions
+        if !&self.to.adjacent_to(&self.from) {
             return Err(SantoriniError::InvalidMovement {
                 from: self.from,
                 to: self.to,
+                details: "The given positions are not adjacent".into(),
             });
         }
+
+        // 5) the `to` space is not occupied by another worker
+        if state.space(&self.to).worker().is_some() {
+            return Err(SantoriniError::InvalidMovement {
+                from: self.from,
+                to: self.to,
+                details: "There is a worker blocking the move".into(),
+            });
+        }
+
+        // 6) `to` space is not Domed
+        if state.space(&self.to).tower().dome().is_some() {
+            return Err(SantoriniError::InvalidMovement {
+                from: self.from,
+                to: self.to,
+                details: "There is a dome blocking the move".into(),
+            });
+        }
+
+        // 7) level of the `to` space <= level of the `from` space + 1
+        let to_level = state.space(&self.to).tower().level();
+        let from_level = state.space(&self.from).tower().level();
+        if to_level > from_level.up() {
+            return Err(SantoriniError::InvalidMovement {
+                from: self.from,
+                to: self.to,
+                details: format!(
+                    "Cannot move from level {} to level {}",
+                    from_level, to_level
+                ),
+            });
+        }
+
+        Ok(())
+    }
+
+    fn execute(&self, state: &mut State) -> Result<(), SantoriniError> {
+        self.can_execute(state)?;
+
+        // Safely unwrap since we check can_execute?
```

Comment:
Tbh I'm not really a big fan of this. Maybe the alternative way to do this whole thing is
1. copy game state
2. Attempt execute
3. in event of failure, rollback to copy